go/types/struct.go

```diff
@@ -106,66 +118,93 @@ func (s Struct) desc() StructDesc {
 // MaybeGet returns the value of a field in the struct. If the struct does not a have a field with
 // the name name then this returns (nil, false).
 func (s Struct) MaybeGet(n string) (Value, bool) {
-	_, i := s.desc().findField(n)
+	f, i := s.findField(n)
 	if i == -1 {
 		return nil, false
 	}
-	return s.values[i], true
+	return f.value, true
+}
+
+func (s Struct) findField(name string) (*structValueField, int) {
```

Comment:
I suggest adding `searchField` which `findField` calls. i.e.

```
func (s Struct) searchField(name string) int {
	return sort.Search(len(s.fieldNames), func(i int) bool { return s.fieldNames[i] >= name })
}

func (s Struct) findField(name string) (idx int, found bool) {
	idx := s.searchField(name)
	if idx == len(s.fields) || s.fieldNames[idx] != name {
		return -1, false
	}
	return idx, true
}
```

that was the mutation functions (set, in particular) below can use `searchField` to find the position that the field goes and use `copy` to efficiently create new fieldName and values slices, without needing to sort. i.e. something like:

```
func (s Struct) Set(n string, v Value) Struct {
	i := s.searchField(n)
	var fieldNames []string
        var values ValueSlice

	if i != len(s.fieldNames) && s.fieldNames[i] == n {
		fieldNames = s.fieldNames
		values = make(ValueSlice, len(s.values))
		copy(values, s.values)
		values[i] = v
	} else {
		fieldNames = make([]string, len(s.fieldNames)+1)
		values = make(ValueSlice, len(s.values)+1)
		copy(values[:i], s.values[:i])
		values[i] = v
		copy(values[i+1:], s.values[i:])
	}

	return Struct{s.name, fieldNames, values, &hash.Hash{}}
}
```
```
func (s Struct) Delete(n string) Struct {
	idx := s.findField(n)
	if idx == -1 {
		return s
	}

	fieldNames := make(string, len(s.values)-1)
	copy(fieldNames[:i], s.fieldNames[:i])
	copy(fieldNames[i-1:], s.fieldNames[:i])

	values := make(string, len(s.values)-1)
	copy(values[:i], s.values[:i])
	copy(values[i:], s.values[i+1:])

	return Struct{s.name, fieldNames, values, &hash.Hash{}}
}
```